cloud/kyup/kyup.py

```diff
@@ -0,0 +1,338 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: kyup
+short_description: create, terminate, start or stop a container in kyup.com, return container_id.
+description:
+    - Creates or terminates Kyup container. When created optionally waits for it to be 'running'. This module has a dependency on python-json and python-requests.
+version_added: "2.2"
+author: Marian Marinov <mm@kyup.com>
+options:
+  api_key:
+    description:
+      - Kyup.com API key. The key can also be supplied by setting the Environment variable KYUP_API_KEY.
+    required: true
+  enc_key:
+    description:
+      - Kyup.com encryption key. The key can also be supplied by setting the Environment variable KYUP_ENC_KEY.
+    required: false
+  action:
+    description:
+      - Action you want to perform on the container.
+    required: true
+  name:
+    description:
+      - Name of the container that you will create/operate on
+    required: true
+  ssh_keys:
+    description:
+      - Comma separated list of ssh key names, that have access to this container.
+    required: false
+  password:
+    description:
+      - Password for the root user of the container. Required when I(action) = 'create'.
+    required: false
+  image:
+    description:
+      - Template image to use when creating the container. Required when I(action) = 'create'.
+    required: false
+  dc_id:
+    description:
+      - DataCenter location ID. Required when I(action) = 'create'.
+    required: false
+    default: 1
+  mem:
+    description:
+      - Memory of the container in GB. Minimum 1 GB.
+    required: false
+    default: 1
+  hdd:
+    description:
+      - Storage size in GB. Minimum 20 GB.
+    required: false
+    default: 20
+  cpu_cores:
+    description:
+      - Number of CPU cores to be used. Minimum 2 CPU cores.
+    required: false
+    default: 2
+  bw:
+    description:
+      - Bandwidth limit for this container, in TB. Minimum 2 TB.
+    required: false
+    default: 2
+  storage_type:
+    description:
+      - Choose the storage type for this container.
+    required: false
+    default: local
+'''
+
+EXAMPLES = '''
+# Basic provisioning example
+---
+- hosts: kyup
+  tasks:
+  - name: ensure kyup container
+    kyup: >
+      api_key=XXXX
+      enc_key=YYYY
+      action=create
+      image="Centos Plain"
+      dc_id=1
+      name=avalon
+      password=e4a17265b28f6a9e8657
+      ssh_keys=hackman,piele
+  - add_host: name={{ container_ip }} group=containers
+
+- hosts: containers
+  tasks:
+  - name: copy the hosts file
+    copy: src=/etc/hosts dest=/etc/hosts
+
+
+# Rebooting a container
+---
+- hosts: kyup
+  tasks:
+  - name: restart the container
+    kyup: >
+      api_key=XXXX
+      action=restart
+      name=avalon
+
+'''
+
+import os
+import requests
```

Comment:
If executed in a python environment without requests installed the module will fail.
For example when this task needs to be delegated to a host with internet access(SNAT) in a controlled environment.